src/libraries/System.Linq/src/System/Linq/Distinct.cs

```diff
@@ -22,7 +22,26 @@ public static IEnumerable<TSource> Distinct<TSource>(this IEnumerable<TSource> s
                 return [];
             }
 
-            return new DistinctIterator<TSource>(source, comparer);
+            return DistinctIterator(source, comparer);
+        }
+
+        private static IEnumerable<TSource> DistinctIterator<TSource>(IEnumerable<TSource> source, IEqualityComparer<TSource>? comparer)
+        {
+            using IEnumerator<TSource> enumerator = source.GetEnumerator();
+
+            if (enumerator.MoveNext())
+            {
+                var set = new HashSet<TSource>(DefaultInternalSetCapacity, comparer);
+                do
+                {
+                    TSource element = enumerator.Current;
```

Comment:
The `IEnumerator` usage is invalid here. An enumerator starts as an invalid state, and the first `MoveNext` will return `false` for an empty one.